[MRG] adding as_frame functionality for california housing dataset loader

gitsteph opened the pull request:
#### Reference Issues/PRs
Handles part of #10733 
(Loosely following conventions found in #13902)

#### What does this implement/fix? Explain your changes.
Adds `as_frame` functionality for the California Housing dataset loader (`fetch_california_housing`)

#### Any other comments?
No

gitsteph committed:
adding as_frame functionality for california housing dataset loader

amueller commented:
the dataframe should also contain the target, right?

gitsteph commented:
Ah, fixing that now

amueller commented:
you can make the change in this PR.

gitsteph commented:
Will do; just noticed a couple additional things to fix so wanted to close it until they're handled.  It's not ready for viewing yet 😅

gitsteph committed:
fixes for as_frame to fetch_california_housing

gitsteph committed:
minor change

gitsteph committed:
forgot to rename feature_names to columns here

gitsteph commented:
@amueller hope this is better 🙏 
thanks for the feedback :)

amueller commented on sklearn/datasets/_california_housing.py:
```diff
@@ -155,10 +171,23 @@ def fetch_california_housing(data_home=None, download_if_missing=True,
+    target_names = ["MedHouseVal", ]
+    if as_frame:
+        columns = feature_names + target_names
+        adjusted_data = np.hstack((data, target[:,np.newaxis]))
```
This is not a good idea in case X and y have different data types. numpy doesn't allow different types, so you're making ``y`` be float here (if it was int) or ``X`` be object (if y is object).

And I think the approach of wrapping this in a reusable function is good.

gitsteph replied:
The current implementation with numpy worked for this specific dataset, but I agree that it wouldn't generalize well. 
 I'll try making this more generalizable -- maybe by constructing separate dataframes (one for `X` and one for `y`), and using pandas to concatenate them together instead of numpy.  🤔

amueller replied:
yes something like that would probably work well. Have you tried working with  @wconnell as well?

gitsteph replied:
Cool!  I just made that proposed change (in the latest commit below) and tested it locally.  The `_convert_data_dataframe` method in this PR hopefully will be able to generalize to some other cases.

I haven't tried working with @wconnell but am open to it!  Will try to find them on gitter.

gitsteph committed:
changes to use pandas concat and make conversion to df more generalizeable

amueller commented on sklearn/datasets/_california_housing.py:
```diff
@@ -155,10 +176,19 @@ def fetch_california_housing(data_home=None, download_if_missing=True,
+    frame = None
+    target_names = ["MedHouseVal", ]
+    if as_frame:
+        frame, X, y = _convert_data_dataframe("fetch_california_housing", data, target, feature_names, target_names)
```
break at 79 chars please

amueller commented:
please add a test that the attribute exists and has the expected content. Otherwise looks good!

amueller commented on sklearn/datasets/_california_housing.py:
```diff
@@ -48,8 +49,17 @@
 logger = logging.getLogger(__name__)
 
 
+def _convert_data_dataframe(caller_name, data, target, feature_names, target_names):
```
this probably should live in ``sklearn/datasets/base.py``

gitsteph replied:
Just moved it after chatting with @wconnell :)
(I left it here temporarily since I wasn't sure it would work for all cases; it's in `_base.py` in the most recent commit)

gitsteph committed:
moved _convert_data_dataframe to _base.py

gitsteph committed:
breaking at 79 chars

wconnell commented:
When we implement this in the various `load_*` functions we will also have to cover the case when `return_X_y=True` and `as_frame=True`.

amueller commented:
Indeed. In this case, the function should (apparently) return X and y separately as a DataFrame and a Series respectively.

gitsteph committed:
added test_fetch_asframe

amueller commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -252,6 +253,18 @@ def test_loads_dumps_bunch():
     assert bunch_from_pkl['x'] == bunch_from_pkl.x
 
 
+def test_fetch_asframe():
```
the test should be in test_california_housing.py I think.

gitsteph replied:
Gotcha; thanks!  Wasn't sure since it aims to test a function in the `_base.py` file.  I'll move it now.

gitsteph committed:
moved test to test_california_housing.py

gitsteph committed:
moved return_X_y handling below as_frame check

gitsteph commented on sklearn/datasets/_california_housing.py:
```diff
@@ -181,6 +179,9 @@ def fetch_california_housing(data_home=None, download_if_missing=True,
                                               feature_names,
                                               target_names)
 
+    if return_X_y:
```
Moved this to better handle the case where both `return_X_y` and `as_frame` are `True`.  If they are both true, this will return `X` and `y` as pandas objects.

gitsteph replied:
TY to @wconnell for this comment => [(https://github.com/scikit-learn/scikit-learn/pull/15486#issuecomment-549087417)]

amueller commented:
test are failing

gitsteph commented:
👌 I'll fix the remaining linting issues now.

🔍 As for the other failed tests -- it seems that the test environments don't have pandas installed (?).  I'm not sure how to test the new functionality in that case... would you have any advice on how to proceed with that?  (e.g. finding a way to install pandas into those test environments, or some other workaround?)

gitsteph committed:
fixed linting issues caught by flake8

gitsteph committed:
using pytest.importorskip to handle test environments without pandas installed

gitsteph committed:
skip test if dataset is not available in test env

gitsteph commented:
I think I've fixed most all of the tests... except `codecov/patch` . I don't quite understand the details of what is failing there.  I understand it's related to testing coverage, but it looks to be picking up "coverage changes" in files that I did not modify -- specifically in `sklearn/manifold/_mds.py`.

None of the files I modified directly are listed there... Not sure how to proceed and would appreciate some guidance for this.  :)

gitsteph commented:
For `test_fetch_asframe` -- I'm failing the coverage test (`codecov/patch`) because I added `@pytest.mark.skipif` in the last commit (https://github.com/scikit-learn/scikit-learn/pull/15486/commits/45940b22d2ee7af6cca9a5764ecdc67a537de2cc), following the pattern of how Olivetti faces dataloading test handled the data not being available at test time.

The original California housing dataloading test (`test_fetch`) also did not run if the data was not available.

I'm not sure the best way to proceed on this.  Any guidance here would be appreciated.  🙏

jnothman commented on sklearn/datasets/tests/test_california_housing.py:
```diff
@@ -13,14 +15,37 @@ def fetch(*args, **kwargs):
+    pd = pytest.importorskip('pandas')
+    bunch = fetch(as_frame=True)
+    frame = bunch.frame
+    assert (hasattr(bunch, 'frame') is True)
```
```suggestion
    assert hasattr(bunch, 'frame') is True
```

gitsteph replied:
Cool; just committed these two suggestions!  Thanks!

jnothman commented on sklearn/datasets/tests/test_california_housing.py:
```diff
@@ -13,14 +15,37 @@ def fetch(*args, **kwargs):
+    bunch = fetch(as_frame=True)
+    frame = bunch.frame
+    assert (hasattr(bunch, 'frame') is True)
+    assert (frame.shape == (20640, 9))
```
```suggestion
    assert frame.shape == (20640, 9)
```

jnothman reviewed (approved):
Otherwise LGTM

gitsteph committed:
Update sklearn/datasets/tests/test_california_housing.py

Removed unneeded parens

Co-Authored-By: Joel Nothman <joel.nothman@gmail.com>

gitsteph committed:
Update sklearn/datasets/tests/test_california_housing.py

Removed unneeded parens

Co-Authored-By: Joel Nothman <joel.nothman@gmail.com>

reshamas commented:
@gitsteph thanks for the email.  I added this PR to the [LIST](https://github.com/WiMLDS/bayarea-2019-scikit-sprint/projects/1)

@cmarmo Adding this PR to the SF sprint list.  

@thomasjpfan Are you able to assist on this one?  Do we know why the check is failing?

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into pr/15486

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into pr/15486

cmarmo commented:
@gitsteph  , the two failing builds are failing because of ubuntu issues:
```
E: Failed to fetch https://packages.microsoft.com/ubuntu/16.04/prod/dists/xenial/main/binary-amd64/Packages.bz2  Hash Sum mismatch
```
Would you please sync with master and push... hope that the issues have been resolved.
Then, maybe @amueller will be able to finalize his review?
